kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,72 @@
+pub trait Client<'a, T: DigestType> {
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// On error or success `data` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a digest is computed.
+    /// On error or success `digest` will contain a reference to the original
+    /// digest supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), ReturnCode>, digest: &'static mut T);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a, T: DigestType> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a, T>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ReturnCode, &'static mut [u8])>;
+
+    /// Request the hardware block to generate a Digest and stores the returned
+    /// digest in the memory location specified.
+    /// This doesn't return any data, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// On error the return value will contain a return code and the original digest
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    ///
+    /// If an appropriote `set_mode*()` wasn't called before this function the
+    /// implementation should try to use a default option. In the case where
+    /// there is only one digest supported this should be used. If there is no
+    /// suitable or obvious default option, the implementation can return an
+    /// error with error code ENOSUPPORT.
+    fn run(&'a self, digest: &'static mut T) -> Result<(), (ReturnCode, &'static mut T)>;
+
+    /// Clear the keys and any other sensitive data.
+    /// This won't clear the buffers provided to this API, that is up to the
+    /// user to clear.
+    fn clear_data(&self);
+}
+
+pub trait HMACSha256 {
+    /// Call before `Digest::run()` to perform HMACSha256
+    ///
+    /// The key used for the HMAC is passed to this function.
+    fn set_mode_hmacsha256(&self, key: &[u8]) -> Result<(), ReturnCode>;
```

Comment:
This should take an array of the keysize. One issue that came up on the OpenTitan call is a concern about hardware that allows you to specify arbitrary length keys that are then hashed. This should be a separate API -- it is not an HMACSha256, it's a KDF followed by HMAC.

Reply:
Fixed

Reply:
I believe this to not be entirely true. HMAC ([RFC 2104, 2. Definition of HMAC](https://tools.ietf.org/html/rfc2104#section-2)) explicitly states that "[the] authentication key K can be of any length up to B, the block length of the hash function". While the padding is well defined (all zeros), this shouldn't be propagated to the client: IMHO to be closely following the standard the trait should accept at least any key length up to `K`, which a fixed length array would make impossible (without an additional length attribute).

Going further, under [3. Keys](https://tools.ietf.org/html/rfc2104#section-3), the reference explicitly states that "The key for HMAC can be of any length (keys longer than B bytes are first hashed using H)", which leads me to believe that accepting a key `> K` is an integral part of a standards compliant HMAC implementation.

Reply:
A good middle ground would be a trait which is standards compliant by allowing any key size, always padding if necessary, but not necessarily hashing keys longer than `K`. If a long key is passed and hashing (the KDF part) is not supported (possibly not conforming to the original standard RFC 2104), return `ReturnCode::ESIZE`, indicating that hashing is required first. This exact required behavior would need to be documented in the trait.

Reply:
What are the thought here? I kind of think a fixed key length makes the most sense as Phil says it means every implementation will do the same thing. To pad with zeros or hash the key (KDF) we should add some other call that makes it explicit. Otherwise every Digest implementation needs to make sure it handles the keys correctly, which they might not. If we split it out we can just have one common place that does the key changes in common code..

Reply:
I think the best approach is to have a fixed-length key, but then you can always layer something more flexible on top. Having a fixed size key allows you to avoid runtime error checking. If the call takes a slice, then you want to return an error if, say, the key is of length 0.  Moving to slice for a key will also require that setting the key has a callback, because computing the hash does. The call stops being synchronous.

So I think the right middle ground is to keep the HMACSha256 as is, but we can have later another trait, HMACSha256Hashed, whose set_key method takes a slice and which issues a callback.

Incidentally, should it be HmacSha256?

Reply:
I do see the advantages a fixed size key has regarding implementation complexity.

I'm also certain having another trait (`HMACSha256Hashed`) will lead to that not being implemented / used at least some of the time. We'd then have a widely accessible kernel-internal (and not crate internal etc.) interface (`HMACSha256`) which sort of implements HMAC, is _called_ HMAC, but really doesn't adhere to standards.

@phil-levis If I understand you correctly, when having a piece of hardware not supporting KDF it would implement `HMACSha256`. HW supporting that would implement `HMACSha256Hashed`. When we now start writing clients using `HMACSha256`, those will be inherently incompatible with `HMACSha256Hashed`, because a HW can't implement _both_ traits as we'd then violate the idea of the single data path in the crypto traits (which still needs discussion). The only solution I'd be comfortable with is immediately writing a wrapper struct for `HMACSha256`s, exposing them as `HMACSha256Hashed` so clients can exclusively use that trait (which would allow us to add HW later that does the KDF implicitly). That relies on a hasher though.

It's a difficult discussion - I am not the one to decide. Deliberately exposing interfaces not adhering to the standards could hurt us in the long run, so I'd like to be careful about that.